### PR TITLE
[`core`] Fix safetensors serialization for shared tensors

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/peft/issues/1079

Safetensors do not support tensor aliasing, before saving the model state dict, we need to loop over the state dict and check for aliased tensors and `clone()` them to avoid any error

cc @BenjaminBossan @pacman100

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -199,6 +203,25 @@ def save_pretrained(
             os.makedirs(output_dir, exist_ok=True)
 
             if safe_serialization:
+                # Safetensors does not allow tensor aliasing.
```
Comment (BenjaminBossan) [inline src/peft/peft_model.py:206]: This block is basically copied from https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L2111-L2134, right? Let's add a comment.

File: src/peft/peft_model.py [hunk 2]
```diff
@@ -199,6 +203,25 @@ def save_pretrained(
...
+                # We're going to remove aliases before saving
+                ptrs = collections.defaultdict(list)
+                for name, tensor in output_state_dict.items():
+                    # Sometimes in the state_dict we have non-tensor objects.
+                    # e.g. in bitsandbytes we have some `str` objects in the state_dict
+                    if isinstance(tensor, torch.Tensor):
+                        ptrs[id_tensor_storage(tensor)].append(name)
+                    else:
+                        # In the non-tensor case, fall back to the pointer of the object itself
+                        ptrs[id(tensor)].append(name)
```
Comment (younesbelkada) [inline src/peft/peft_model.py:216]: Not sure if this is needed, I added that in the transformers code to handle bnb state dict case and should be safe to keep it but happy to remove it as well

Comment (BenjaminBossan) [reply src/peft/peft_model.py:216]: Probably it's better to follow the same steps as transformers.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for investigating this. Just a minor comment about adding a comment, apart from this, LGTM.

File: src/peft/peft_model.py (hunk 1 above)
Comment (younesbelkada) [reply src/peft/peft_model.py:206]: Ok perfect!

**Pull request merged**